### ISSUE TITLE
gradient inconsistency problem

### ISSUE THREAD

**Issue by zcyang** [cvxgrp/cvxpylayers#70]
Sorry I cannot stop digging into the gradient problem and want to figure out whether it is a bug or a computation precision problem. Wonder if there is a quick fix. Thanks!

This may be related to previous issues of #69 and #67 

Suppose the loss is composed of two parts `l = l1 + l2`, then the gradient should following the same addition rule:
`$\delta l_w = \delta l1_w + \delta l2_w`. However, the following example shows a significant gap between them:

```python
import numpy as np
import torch
# torch.set_default_dtype(torch.double)

import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.autograd import Variable
from torch.nn.parameter import Parameter

import cvxpy as cp
from cvxpylayers.torch import CvxpyLayer

N = 200
M = 100
BOUND = 0.1
NORM_BOUND = 10
NUM_STEPS = 3

class Opt(nn.Module):
  def __init__(self, weight=5.0):
    super(Opt, self).__init__()
    self.weight = Parameter(
      torch.tensor([[weight]], dtype=torch.double, requires_grad=True))
    self.L = torch.randn(N, M)
    L_diag = torch.rand(N)
    L_diag[:N//2] *= 0.1
    L_diag[N//2:] *= 10
    self.L_diag = torch.diag(L_diag)

    self.L = self.L.double()
    self.L_diag = self.L_diag.double()

  def forward(self, x):
    num_steps = x.shape[0]
    xw = torch.matmul(x, self.weight).squeeze()
    init_v = torch.zeros(x.shape[1])
    init_v = init_v.double()
    v_list = []
    for i in range(num_steps):
      _p = cp.Parameter((N))
      _prev = cp.Parameter((N))
      _v = cp.Variable((N))
      obj = cp.Minimize(0.5*cp.sum_squares(self.L.T@_v)
                        + 0.5*cp.sum_squares(self.L_diag@_v)
                        + 50*cp.sum_squares(_v - _prev)
                        - _v@_p)
     
      cons = [_v <= 0.1,
              -_v <= 0.1,
              cp.norm(_v, 1) <= 10
      ]
      prob = cp.Problem(obj, cons)

      _prev.value = init_v.detach().numpy()
      _p.value = xw[i].detach().numpy()

      prob.solve(solver=cp.ECOS)
      v_ecos = _v.value
      if prob.status != "optimal":
        print(prob.status)

      self.cvxopt = CvxpyLayer(prob, [_p, _prev], [_v])

      v, = self.cvxopt(xw[i], init_v, solver_args={"eps": 1e-6, "max_iters": 200000})

      # print("ecos gap abs {:.6f} ".format(
      #   np.max(np.abs(v_ecos - v.detach().numpy())),
      #   ))

      init_v = v
      v_list.append(v)
    v = torch.stack(v_list)

    return v

def main():
  torch.manual_seed(0)
  opt = Opt(weight=5.0)

  x = torch.randn(N, 1)
  x = x.unsqueeze(0).repeat(NUM_STEPS, 1, 1)
  y = 0.1*x[-1] + torch.randn(N, 1)

  x, y = x.double(), y.double()

  optimizer = optim.Adam(opt.parameters(), 1e-1)
  for i in range(1):
    v = opt(x)
    loss1 = -torch.sum(y*v[-1])
    loss2 = torch.sum(v**2)*5

    loss = loss1
    optimizer.zero_grad()
    loss.backward(retain_graph=True)
    # optimizer.step()
    l = loss.detach().numpy().tolist()
    w = opt.weight.detach().numpy()
    g = opt.weight.grad.detach().numpy()
    g1 = g.copy()
    print("loss1 {:.3f} weight {:.3f} "
          "grad {:.3f} ".format(
            l, w[0][0], g[0][0],
          ))

    loss = loss2
    optimizer.zero_grad()
    loss.backward(retain_graph=True)
    # optimizer.step()
    l = loss.detach().numpy().tolist()
    w = opt.weight.detach().numpy()
    g = opt.weight.grad.detach().numpy()
    print("loss2 {:.3f} weight {:.3f} "
          "grad {:.3f} ".format(
            l, w[0][0], g[0][0],
          ))
    g2 = g.copy()

    loss = loss1 + loss2
    optimizer.zero_grad()
    loss.backward()
    # optimizer.step()
    l = loss.detach().numpy().tolist()
    w = opt.weight.detach().numpy()
    g = opt.weight.grad.detach().numpy()
    print("loss {:.3f} weight {:.3f} "
          "grad {:.3f} ".format(
            l, w[0][0], g[0][0],
          ))
    print("gap between grad {:.3f}".format(g[0][0]-(g1[0][0]+g2[0][0])))

```

Output:
```
loss1 -8.689 weight 5.000 grad 0.563
loss2 8.213 weight 5.000 grad 1.798
loss -0.476 weight 5.000 grad 1.999
gap between grad -0.362
```

**Comment by zcyang** [comment]
@bamos  Below is output after setting `solver_args['mode'] = 'dense'`

```
loss1 -8.689 weight 5.000 grad -0.070
loss2 8.213 weight 5.000 grad 1.227
loss -0.476 weight 5.000 grad 1.157
gap between grad 0.000
```

The gap becomes 0.
